### PR TITLE
Make requests emit metrics when rate limited

### PR DESCRIPTION
**Pull request by andrew4699**
# Description

Makes requests emit metrics when they get rate limited.

There were 2 issues preventing this:
- RateLimiterFilter ran outside the Jersey Servlet, which is before application events get sent
- Changing it to run inside the servlet isn't enough. Aborted requests don't emit the RESOURCE_METHOD_START event.

This PR makes the rate limiting filter a Jersey Servlet filter and makes the per-request metric emitter emit count metrics for requests that have a REQUEST_MATCHED But not RESOURCE_METHOD_START event.

## Type of change

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [X] See the test that verifies emitted metrics

# Checklist:

Please delete options that are not relevant.

- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] If adding new functionality, I have discussed my implementation with the community using the linked GitHub issue

### CODE REVIEW

File: polaris-service/src/main/java/org/apache/polaris/service/ratelimiter/RateLimiterFilter.java [hunk 1]
```diff
@@ -18,22 +18,17 @@
...
+import jakarta.ws.rs.container.ContainerRequestContext;
+import jakarta.ws.rs.container.ContainerRequestFilter;
 import jakarta.ws.rs.core.Response;
 import java.io.IOException;
+import javax.ws.rs.ext.Provider;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /** Request filter that returns a 429 Too Many Requests if the rate limiter says so */
-@Priority(Priorities.AUTHORIZATION + 1)
```
Comment (andrew4699) [inline polaris-service/src/main/java/org/apache/polaris/service/ratelimiter/RateLimiterFilter.java:35]: Fun fact: this actually did nothing previously. It's a lucky coincidence that this ran after the ContextResolverFilter due to addFilter ordering.

Comment (eric-maynard) [reply polaris-service/src/main/java/org/apache/polaris/service/ratelimiter/RateLimiterFilter.java:35]: Removing this moves it back to USER right?

File: polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java [hunk 2]
```diff
@@ -261,10 +261,7 @@ public void run(PolarisApplicationConfig configuration, Environment environment)
         .addMappingForUrlPatterns(EnumSet.of(DispatcherType.REQUEST), true, "/*");
 
     if (configuration.getRateLimiter() != null) {
-      environment
-          .servlets()
-          .addFilter("ratelimiter", new RateLimiterFilter(configuration.getRateLimiter()))
-          .addMappingForUrlPatterns(EnumSet.of(DispatcherType.REQUEST), true, "/*");
+      environment.jersey().register(new RateLimiterFilter(configuration.getRateLimiter()));
```
Comment (eric-maynard) [inline polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:264]: This seems like a semantic change unrelated to the metrics. If I'm reading this correctly, this changes the rate limiter to only apply to Jersey resources

Comment (jbonofre) [inline polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:264]: This change is unrelated to metrics, right ?

File: polaris-service/src/main/java/org/apache/polaris/service/TimedApplicationEventListener.java [hunk 3]
```diff
@@ -66,16 +66,16 @@ private class TimedRequestEventListener implements RequestEventListener {
     @Override
     public void onEvent(RequestEvent event) {
       String realmId = CallContext.getCurrentContext().getRealmContext().getRealmIdentifier();
-      if (event.getType() == RequestEvent.Type.RESOURCE_METHOD_START) {
+      if (event.getType() == RequestEvent.Type.REQUEST_MATCHED) {
```
Comment (jbonofre) [inline polaris-service/src/main/java/org/apache/polaris/service/TimedApplicationEventListener.java:69]: FYI, this listener to `@TimedApi` (populated in mustache template) is not required with Quarkus.

File: polaris-service/src/main/java/org/apache/polaris/service/ratelimiter/RateLimiterFilter.java (hunk 1 above)
Comment (andrew4699) [reply polaris-service/src/main/java/org/apache/polaris/service/ratelimiter/RateLimiterFilter.java:35]: Yes. Previously I don't think it did anything because it was outside Jersey.

File: polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java (hunk 2 above)
Comment (andrew4699) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:264]: Yes because TimedApplicationEventListener, which is what emits metrics on requests, depends on Jersey events.

Comment (andrew4699) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:264]: See https://github.com/apache/polaris/pull/423#discussion_r1828923405

Comment (eric-maynard) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:264]: I get that, but my concern is whether this actually changes the behavior of the system by rate-limiting less things now

Comment (andrew4699) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:264]: (Please fact check me as I only started working with Java web services recently)

I think all of our endpoints are in Jersey, so it shouldn't change the scope, but at a minimum it does push rate limiting further into the request handling chain.

The fundamental problem seems to be that only Jersey knows how to map a request path to its handling method (and thus its metric name). If that mapping was available everywhere, TimedApplicationEventListener could be re-written as a top level filter.

File: polaris-service/src/test/java/org/apache/polaris/service/ratelimiter/RateLimiterFilterTest.java [hunk 4]
```diff
@@ -70,20 +75,58 @@ public static void setup(
...
     requestAsserter.accept(Response.Status.TOO_MANY_REQUESTS);
 
-    clock.add(Duration.ofSeconds(4 * WINDOW_SECONDS)); // Clear any counters from during this test
+    assertTrue(
+        getCounter(
+                "polaris_principal_roles_listPrincipalRoles_error_total{HTTP_RESPONSE_CODE=\"429\"}")
+            > 0);
+  }
+
+  private double getCounter(String metricName) {
```
Comment (andrew4699) [inline polaris-service/src/test/java/org/apache/polaris/service/ratelimiter/RateLimiterFilterTest.java:117]: Will replace this with https://github.com/apache/polaris/blob/cfd908ad3bc7527d5fbd2127f8eb7bce54a0acbd/polaris-service/src/test/java/org/apache/polaris/service/test/TestMetricsUtil.java once my other PR is merged

File: polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java (hunk 2 above)
Comment (collado-mike) [reply polaris-service/src/main/java/org/apache/polaris/service/PolarisApplication.java:264]: Semantically, there is a difference, as ServletFilters are applied before Jersey Filters. Practically, however, it shouldn't have an impact. I think the only real dependency the RateLimiter might have is that the RealmContext is set in case there are different rate limits per realm. Since the `ContextResolverFilter` still applies first, we shouldn't see a practical difference.

**Review by collado-mike — APPROVED** [review]
